elastic/apm-server
# Add a reason to some auth failures

## Motivation/summary

Add a reason to some auth failures, to better inform users of how to configure clients. The reasons we include are limited to:
 - invalid authorization kind, e.g. "Authorization: Bear token" instead of "Authorization: Bearer token"
 - missing or improperly formatted Authorization header, e.g. "Authorization: token" instead of "Authorization: Bearer token"

Most importantly, no details of the server's configuration are leaked. The auth failure reason is used to populate the error message sent back to clients, in both HTTP and gRPC requests.

There is also a change in this PR to return 503 Service Unavailable rather than 401 Unauthorized when an auth implementation returns an error. For example if Elasticsearch cannot be reached for API Key auth, that's a 503 and not 401. This addresses #5061.

Finally, there is some refactoring in this PR to make this all possible:
 - We introduce a `beater/authorization.Result` struct type, which holds a boolean indicating whether auth was successful, and an optional reason if not.
 - We no longer store a `beater/authorization.Authorization` on the request object, but instead store a Result.
 - The Authorization interface has been simplified, and the Deny/AllowAuth implementations have been unexported.

Later on we could look at extending the Result struct for storing auth-related information in documents as described in https://github.com/elastic/apm-server/issues/3731.

## Checklist


## How to test these changes

1. Configure the server with a secret token, say "abc123"
2. Using curl, send a POST request to `/intake/v2/events` with:
  - no Authorization header: server should respond saying header is missing or badly formatted
  - a badly formatted Authorization header (`-H "Authorization: abc123"`): server should respond saying header is missing or badly formatted
  - an Authorization header with an invalid auth kind (`-H "Authorization: Secret Token"`): server should respond saying "Secret" is not a valid auth kind
  - an Authorization header with an invalid secret token (`-H "Authorization: Bearer Token"`): server should respond `unauthorized`
3. Run a subset of these tests an OpenTelemetry SDK, proving the OTLP/gRPC intake is also covered.

## Related issues

Closes #4880 
Closes #5061
